### PR TITLE
Shaders for Android (GLES 2)

### PR DESCRIPTION
**Pull request by numberZero**
Adds shader support for OpenGL ES 2 device (used on Android). Works for me™
photo5219674515494252531
Based on #7514

### CODE REVIEW

File: src/config.h [hunk 1]
```diff
@@ -16,6 +16,7 @@
 	#define PROJECT_NAME_C "Minetest"
 	#define STATIC_SHAREDIR ""
 	#define VERSION_STRING STR(VERSION_MAJOR) "." STR(VERSION_MINOR) "." STR(VERSION_PATCH) STR(VERSION_EXTRA)
+	#define ENABLE_GLES 1
```
Comment (MoNTE48) [inline src/config.h:19]: Shouldn't it get into PR?

File: src/client/shader.cpp [hunk 2]
```diff
@@ -620,15 +653,58 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
...
+				"attribute vec4 inVertexColor;\n" \
+				"attribute vec4 inTexCoord0;\n" \
+	/* 			"attribute vec4 inTexCoord1;\n" */\
+				"attribute vec4 inVertexTangent;\n" \
+				"attribute vec4 inVertexBinormal;\n" \
+				"\n" \
+				"#define gl_ModelViewMatrix mWorldView\n" \
+				"#define gl_TextureMatrix mTexture\n" \
+				"#define gl_NormalMatrix mNormal\n" \
+				"#define gl_Vertex inVertexPosition\n" \
```
Comment (sfan5) [inline src/client/shader.cpp:682]: All the `\`'s are not necessary here

File: src/client/shader.cpp [hunk 3]
```diff
@@ -620,15 +653,58 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
...
+			"precision mediump float;\n" \
+			;
+	} else {
+		shaders_header = \
+			"#version 120\n" \
+			;
+	}
+
+
+	bool use_discard = use_gles;
```
Comment (sfan5) [inline src/client/shader.cpp:699]: why is this turned on by default?

File: src/client/shader.cpp [hunk 4]
```diff
@@ -748,19 +824,25 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
...
 	shaders_header += "\n";
 
 	// Call addHighLevelShaderMaterial() or addShaderMaterial()
 	const c8* vertex_program_ptr = 0;
 	const c8* pixel_program_ptr = 0;
 	const c8* geometry_program_ptr = 0;
 	if (!vertex_program.empty()) {
-		vertex_program = shaders_header + vertex_program;
+		if (use_gles)
+			vertex_program = shaders_header + vertex_header + vertex_program;
```
Comment (sfan5) [inline src/client/shader.cpp:836]: Isn't vertex_header left empty anyway? No need for an `if`

File: src/cmake_config.h.in [hunk 5]
```diff
@@ -14,6 +14,7 @@
 #define STATIC_LOCALEDIR "@LOCALEDIR@"
 #define BUILD_TYPE "@CMAKE_BUILD_TYPE@"
 #define ICON_DIR "@ICONDIR@"
+#cmakedefine01 ENABLE_GLES
```
Comment (sfan5) [inline src/cmake_config.h.in:17]: duplicate, see line 31

File: src/config.h (hunk 1 above)
Comment (sfan5) [reply src/config.h:19]: Yeah this looks like it shouldn't be here.

**Comment by lhofhansl** [comment]
This will need a major rebase now after #10487 . Sorry :(

File: src/client/shader.cpp (hunk 3 above)
Comment (numberZero) [reply src/client/shader.cpp:699]: because there is no alpha testing at all in GLES 2

File: src/client/shader.cpp (hunk 4 above)
Comment (numberZero) [reply src/client/shader.cpp:836]: ugh, really. thanks for spotting

File: src/cmake_config.h.in (hunk 5 above)
Comment (numberZero) [reply src/cmake_config.h.in:17]: hmm, yes. that’s probably due to rebase

File: src/client/shader.cpp (hunk 2 above)
Comment (numberZero) [reply src/client/shader.cpp:682]: and you don’t mind redefining the reserved words...

**Comment by numberZero** [comment]
Will rebase (and squash probably). Some cleanup is still necessary, in addition to what was mentioned in the comments.

File: src/client/shader.cpp (hunk 2 above)
Comment (sfan5) [reply src/client/shader.cpp]: those aren't reserved in GLES's GLSL are they?

File: src/client/shader.cpp (hunk 3 above)
Comment (sfan5) [reply src/client/shader.cpp:701]: This was specifically added in #10036 because it was only broken on *one* GPU, so that seems hard to believe.

Comment (numberZero) [reply src/client/shader.cpp:703]: It was also broken in GLES 2 in each attempt of enabling shaders. Read the [OGLES2 reference](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/), there are no alpha test-related functions (glAlphaFunc, GL_ALPHA_TEST in glEnable). And IrrLicht built-in shaders emulate the test too, take a look at `COGLES2TransparentAlphaChannelRef.fsh` in the IrrLicht shaders.

File: src/client/shader.cpp (hunk 2 above)
Comment (numberZero) [reply src/client/shader.cpp]: >  Identifiers starting with “gl_” are reserved for use by OpenGL ES.  No user-defined identifiers may begin with “gl_”.

**Comment by numberZero** [comment]
On my device it works equally well with GLES 1 and GLES 2+Minetest shaders. GLES 2+IrrLicht shaders is considerably slower.

File: src/client/shader.cpp (hunk 3 above)
Comment (sfan5) [reply src/client/shader.cpp:703]: Ah yea, GLES previously used Irrlicht's builtin shaders (emulating the a fixed pipeline) so of course this didn't make a difference.

File: src/client/shader.cpp [hunk 6]
```diff
@@ -620,15 +653,62 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
...
+			)";
+		pixel_header = R"(
+			precision mediump float;
+			)";
+	} else {
+		shaders_header = R"(
+			#version 120
+			#define lowp
+			#define mediump
+			#define highp
```
Comment (sfan5) [inline src/client/shader.cpp:686]: https://www.khronos.org/opengl/wiki/Type_Qualifier_(GLSL)#Precision_qualifiers

According to this lowp/mediump/highp are valid but don't have any meaning, so no need to define them.

File: builtin/settingtypes.txt [hunk 7]
```diff
@@ -658,8 +658,9 @@ texture_path (Texture path) path
 #    The rendering back-end for Irrlicht.
 #    A restart is required after changing this.
 #    Note: On Android, stick with OGLES1 if unsure! App may fail to start otherwise.
-#    On other platforms, OpenGL is recommended, and it’s the only driver with
-#    shader support currently.
+#    On other platforms, OpenGL is recommended.
+#    Shaders are supported by OpenGL (desktop only) and OGLES2 (experimental,
+#    Android only)
```
Comment (sfan5) [inline builtin/settingtypes.txt:663]:
```suggestion
#    Shaders are supported by OpenGL (desktop only) and OGLES2 (experimental)
```

File: client/shaders/nodes_shader/opengl_fragment.glsl [hunk 8]
```diff
@@ -56,10 +57,9 @@ vec4 applyToneMapping(vec4 color)
...
 {
 	vec3 color;
-	vec2 uv = gl_TexCoord[0].st;
+	vec2 uv = varTexCoord.st;
 
 	vec4 base = texture2D(baseTexture, uv).rgba;
-
 #ifdef USE_DISCARD
 	// If alpha is zero, we can just discard the pixel. This fixes transparency
 	// on GPUs like GC7000L, where GL_ALPHA_TEST is not implemented in mesa.
```
Comment (sfan5) [inline client/shaders/nodes_shader/opengl_fragment.glsl:65]: Should probably be a comment here that says that discard is needed for normal GLES2 too.

File: src/client/shader.cpp (hunk 6 above)
Comment (numberZero) [reply src/client/shader.cpp:686]: They are valid in GLSL 1.4 (OpenGL 3) but MT uses GLSL 1.2 (OpenGL 2). They are reserved and not valid there.

Comment (numberZero) [reply src/client/shader.cpp:686]: Argh, they are reserved... so I shouldn’t define them either, according to the spec. Things become hairy...

File: src/client/shader.cpp [hunk 9]
```diff
@@ -620,15 +653,62 @@ ShaderInfo generate_shader(const std::string &name, u8 material_type, u8 drawtyp
...
 #ifdef __unix__
 	// For renderers that should use discard instead of GL_ALPHA_TEST
 	const char* gl_renderer = (const char*)glGetString(GL_RENDERER);
-	if (strstr(gl_renderer, "GC7000")) {
-		shaders_header += "#define USE_DISCARD\n";
-	}
+	if (strstr(gl_renderer, "GC7000"))
+		use_discard = true;
 #endif
+	if (use_discard)
```
Comment (sfan5) [inline src/client/shader.cpp:710]:
```suggestion
	if (use_discard && shaderinfo.base_material != video::EMT_SOLID)
```

File: builtin/settingtypes.txt (hunk 7 above)
Comment (numberZero) [reply builtin/settingtypes.txt:663]: Does that mean GLES2 is supported on desktop?

**Review by sfan5 — APPROVED** [review]
Tested on "OpenGL ES 3.2 NVIDIA 450.66", works.
:+1: if comments addressed

File: client/shaders/nodes_shader/opengl_fragment.glsl (hunk 8 above)
Comment (numberZero) [reply client/shaders/nodes_shader/opengl_fragment.glsl:65]: Okay.

File: src/client/shader.cpp (hunk 6 above)
Comment (sfan5) [reply src/client/shader.cpp:686]: Oops, didn't check the version. This is a problem indeed.

If that doesn't work out the code could fall back to text substitution (or compiling with `#version 140` if supported?).

Comment (numberZero) [reply src/client/shader.cpp:686]: It works on my system ofc (and doesn’t work without the defines) though, so this compiler (mesa) is sane.

**Review by SmallJoker — APPROVED** [review]
Does not break regular OpenGL. Looks good.

**Pull request merged**